driver-core/src/main/com/mongodb/internal/connection/tlschannel/async/AsynchronousTlsChannelGroup.java

```diff
@@ -425,7 +424,9 @@ private void loop() {
         } catch (Throwable e) {
             LOGGER.error("error in selector loop", e);
         } finally {
-            executor.shutdown();
+            if (!executorIsExternal) {
+                executor.shutdown();
+            }
```

Comment:
I was unable to get a test working that would hit this path.

Reply:
That's weird and is worth figuring out:

- the `finally` block should be executed sometime after calling `AsynchronousTlsChannelGroup.shutdown`
- `AsynchronousTlsChannelGroup.shutdown` is called by `TlsChannelStreamFactoryFactory.close`
- `TlsChannelStreamFactoryFactory.close` is called by `com.mongodb.reactivestreams.client.internal.MongoClientImpl.close`

That is, when we close `com.mongodb.reactivestreams.client.MongoClient`, the highlighted `if` check should be done. I am not sure, though, how would you check that the check was done other than by looking at the code coverage report, or by debugging manually.

Reply:
Should work now.